### PR TITLE
[3.x] Fixed rotate_y property of particle shaders

### LINKED ISSUE #33357: 3D particles initial angle with flag_rotate_y shows erratic behaviour
**Godot version:**
master dc114fa2e

**OS/device including version:**
Windows 10 64bit - 1903 

**Issue description:**
Currently, 3D mesh particles show unexpected behaviour with `flag_rotate_y`, they spin around like crazy with angle values over 20 (from the 0-720 range). Without any angular velocity set in the process material.

particles

With `flag_disable_z` the particles behave like expected, one can set the initial angle at the lifetime start of the particle and then also tune the angular velocity if desired.

As per documentation, the angle is meant as an initial rotation, applied in degrees. `flag_rotate_y` suggests, that 3D particles get an initial rotation applied in the Y axis. Instead, they constantly spin around. 

**Possible solution:**
I believe this could be fixed by applying the rotation matrix only once in the restart block.
https://github.com/godotengine/godot/blob/dc114fa2ef336646f56d71322ba0236c00f8228e/scene/resources/particles_material.cpp#L566

**Minimal reproduction project:**
[ParticleTest.zip](https://github.com/godotengine/godot/files/3808151/ParticleTest.zip)

### PR DESCRIPTION
**Pull request by QbieShay**
Rotate_y flag had an issue which made it almost unusable (see #33357).
I'd like some testing on this one, I haven't found any issue with it, but it'd help if someone else can take a look.

This breaks compat because in order to achieve a random, stable, y rotation, one had to pick super small values in the angle property.

Closes #33357

### CODE REVIEW

File: doc/classes/TileMap.xml
Comment (akien-mga) [inline doc/classes/TileMap.xml:329]: Your commit was wrongly rebased into another unrelated commit it seems.

**Comment by QbieShay** [comment]
@akien-mga darn you were faster to notice my git spaghetti than i was to clean it up. Sorry!

File: scene/resources/particles_material.cpp
Comment (akien-mga) [inline scene/resources/particles_material.cpp:571]: Indentation issue. You may want to set up clang-format / pre-commit hooks to avoid this in future commits :)
https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html

**Comment by akien-mga** [comment]
That bugfix should also be made in the `master` branch as it seems to have the same code here:
```
scene/resources/particles_material.cpp
603:            if (particle_flags[PARTICLE_FLAG_ROTATE_Y]) {
```

**Review by akien-mga — APPROVED** [review]
Looks good to me. Would like to have a quick review from rendering folks if possible, but I think it's fine to include in 3.2.4 even if it changes the behavior (any bugfix does if one relied on the buggy behavior).

**Comment by QbieShay** [comment]
@akien-mga I'm not in favor of merging for 3.2.4, this is a good fix, but it would break a bunch of things (even some of the particles that I made for the tps demo)

In order to obtain random y rotation before, one had to select rotate y and then set an angle of ~0.01 degrees. This would give you a nice, almost static, random rotation. But with this fix, all the particles will be aligned in the same direction, because a 0.01 angle is very small ..

**Comment by lawnjelly** [comment]
Codewise it looks fine, but the question seems whether the particles *should* behave in that way and I'm not at all familiar with particles so can't help on that currently (it's going to be a bit hectic till the RC is released). @clayjohn may be more familiar with particle behaviour?

And as @QbieShay says maybe this fix will break something else, so maybe we need to fix several things at once.

**Comment by QbieShay** [comment]
@lawnjelly the way that the parameter works right now is completely broken.

The rotation is stored in CUSTOM.x, so the rotation of the matrix should be set directly from the CUSTOM.x value. What happens instead is that the TRANSFORM is multiplied by the rotation matrix, which means that if in your custom.x you have accumulated 90degress, at frame1 you will have 90deg, frame2 you will have 180, frame3 270, frame4 360. The addition of velocity is all handled in the way that rotation is added to the CUSOTM.x component, which should just be plainly applied as-is to the transform.

**Comment by lawnjelly** [comment]
Ah I see that makes lots of sense. So it is the right fix but we might have to fix some other bits of the engine that relied on the buggy previous behaviour? Or was it just the demos? If it is just demos they should be ok to break hehe :smiling_imp: !

**Comment by akien-mga** [comment]
> I'm not in favor of merging for 3.2.4, this is a good fix, but it would break a bunch of things (even some of the particles that I made for the tps demo)

But if we can't merge it in 3.2.4, we're unlikely to merge it in the `3.2` branch at all since it will be the same problem for future 3.2.x releases. Or we'd have to add another flag in the `3.2` branch to make the fix opt-in and preserve the previous, buggy behavior.

But yeah we could also merge it for 3.2.5 beta 1, see if anyone complains about their particles having changed behavior, and then assess if we want to add compatibility code.

**Comment by QbieShay** [comment]
What I meant is that we're already in RC stage, so I'd rather give people time to fix their particles with a beta, rather than dropiing the bomb like this :P again I don't think that many people use the property because it's very broken, but who knows :(

**Comment by lentsius-bark** [comment]
@QbieShay I can confirm that I've avoided the property due to its behaviour.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
